File: python/data_structures/bst.py
"""custom implementation of a binary search tree with the purpose of practice
the ADT contains the following methods:
    - insert: insert a new node in the tree
    - search: return the node with the given value
    - delete: delete the given value from the tree
    - successor: return the next in order successor
"""
class BST():
    class Node():
        """Node basic chainable storage unit
        """
        def __init__(self, x=None):
            self.value = x
            self.left = None
            self.right = None
        def __repr__(self):
            return repr(self.value)

    def __init__(self, x=[]):
        self.root = None
        for e in x:
            self.insert(e)

    def __repr__(self):
        res = []
        def r(current_node, level=0):
            if not current_node: return
            r(current_node.left, level+1)
            res.append('\t'*level + f'-->({current_node.value})')
            r(current_node.right, level+1)
        r(self.root)
        return '\n'.join(res)

    def insert(self, x):
        """insert a new node into the bst
        """
        # special case: empty bst
        if not self.root: self.root = self.Node(x); return
        # general case
        def r(current_node, x):
            if current_node.value == x: raise ValueError('No duplicates allowed!')
            if x < current_node.value:
                if current_node.left:
                    r(current_node.left, x)
                else:
                    current_node.left = self.Node(x)
            else:
                if current_node.right:
                    r(current_node.right, x)
                else:
                    current_node.right = self.Node(x)
            return
        return r(self.root, x)

    def search(self, value):
        """standard binary search in the tree
        """
        if not self.root: raise ValueError(f'{value} not found')
        def r(node):
            if not node: raise ValueError(f'{value} not found')
            if value == node.value: return node
            if value < node.value: return r(node.left)
            else: return r(node.right)
        return r(self.root)

    def successor(self, value):
        """get the next in-order successor
        """
        if not self.root or value == None: raise ValueError(f'successor of {value} not found')
        # execute standard binary search... but keeping ranges
        def get(node, r):
            if not node: raise KeyError(f'{value} not in the tree')
            if node.value == value: return (node, r)
            if value < node.value: return get(node.left, node)
            else: return get(node.right, r)
        node, r = get(self.root, None)
        # no right subtree, return the previous bigger ancestor
        if not node.right: return r
        # go all left in the right subtree to get the successor
        node = node.right
        while node.left: node = node.left
        return node
    
    def rotate_right(self, node):
        """standard right rotation
        """
        tmp = self.Node(node.value)
        tmp.left, tmp.right = node.left.right, node.right
        # rotate...
        node.value = node.left.value
        node.left, node.right = node.left.left, tmp

    def rotate_left(self, node):
        """standard left rotation
        """
        tmp = self.Node(node.value)
        tmp.left, tmp.right = node.left, node.right.left
        # rotate...
        node.value = node.right.value
        node.left, node.right = tmp, node.right.right

    def delete(self, value):
        """search for value, and rotate until leaf, then delete
        """
        if value == None or not self.root: raise KeyError(f'{value} not found')
        # standard binary search but keep the parent
        def get(node, parent):
            if not node: raise KeyError(f'{value} not found')
            if node.value == value: return (node, parent)
            if value < node.value: return get(node.left, node)
            else: return get(node.right, node)
        node, parent = get(self.root, None)
        # bring it down to leaf... then is trivial to erase
        def r(node, parent):
            # are you root?
            if not node.left and not node.right:
                if not parent: self.root = None
                if node == parent.right: parent.right = None; return
                else: parent.left = None; return
            # rotate until leaf...
            child = node.left or node.right
            if child == node.left:
                self.rotate_right(node); r(node.right, node)
            else:
                self.rotate_left(node); r(node.left, node)
        r(node, parent)

    def build_from_array(self, indexes):
        """build from an heap like array...
        """
        root = self.root
        def build(node, i, indexes, level=0):
            # aux functions
            left = lambda: i*2 +1
            right = lambda: i*2 +2
            # actual tree traversal
            if left() < len(indexes) and indexes[left()] != -1:
                left_node = Node(indexes[left()])
                node.left = left_node
                build(left_node, left(), indexes, level+1)
            if right() < len(indexes) and indexes[right()] != -1:
                right_node = Node(indexes[right()])
                node.right = right_node
                build(right_node, right(), indexes, level+1)
        build(root, 0, indexes, 0)
        return root

    def traverse_array(self, indexes):
        """method for debugging the heap-like array
        """
        def r(i, indexes, level=0):
            left = lambda: i*2 +1
            right = lambda: i*2 +2
            if left() < len(indexes): r(left(), indexes, level+1)
            if indexes[i] != -1: print('\t'*level, f'-->({indexes[i]})')
            if right() < len(indexes): r(right(), indexes, level+1)
        return r(0, indexes, level=0)

File: cpp/src/bst/src/bst.cpp
#include "bst/bst.h"

namespace bst
{

template<typename T>
Node<T>::Node(T _value): value(_value) {}

template<typename T>
BST<T>::BST() {}

template<typename T>
BST<T>::BST(const std::vector<T>& data) {
	for(T e: data)
		this->insert(e);
}

template<typename T>
void BST<T>::insert(const T& key) {
	if(!this->root) {
		this->root = std::shared_ptr<Node<T>>(new Node<T>(key));
		return;
	}
	return this->insert(key, this->root);
}

template<typename T>
void BST<T>::insert(const T& key, std::shared_ptr<Node<T>> node) {
	//base case: no node
	if(!node)
		return;
	if(node->value == key)
		throw std::invalid_argument("No, duplicates allowed!");
	if(key < node->value) {
		if(node->left)
			this->insert(key, node->left);
		else
			node->left = std::shared_ptr<Node<T>>(new Node<T>(key));
	}
	else {
		if(node->right)
			this->insert(key, node->right);
		else
			node->right = std::shared_ptr<Node<T>>(new Node<T>(key));
	}
}

template<typename T>
std::shared_ptr<Node<T>> BST<T>::max() {
	if(!this->root)
		return nullptr;
	auto node = this->root;
	while(node->right)
		node = node->right;
	return node;
}

template<typename T>
std::shared_ptr<Node<T>> BST<T>::min() {
	if(!this->root)
		return nullptr;
	auto node = this->root;
	while(node->left)
		node = node->left;
	return node;
}

template<typename T>
std::shared_ptr<Node<T>> BST<T>::search(const T& key) {
	if(!this->root)
		return nullptr;
	return this->search(key, this->root);
}

template<typename T>
std::shared_ptr<Node<T>> BST<T>::search(const T& key, std::shared_ptr<Node<T>> node) {
	if(!node)
		return nullptr;
	if(key == node->value)
		return node;
	if(key < node->value)
		return this->search(key, node->left);
	else
		return this->search(key, node->right);
}

template<typename T>
void BST<T>::traverse() {
	if(!this->root)
		return;
	return traverse(this->root);
}

template<typename T>
void BST<T>::traverse(std::shared_ptr<Node<T>> node, int level) {
	if(!node)
		return;
	this->traverse(node->left, level+1);
	for(auto i = 0; i < level; i++)
		std::cout << "\t";
	std::cout << "-->(" << std::to_string(node->value) << ")" << std::endl;
	this->traverse(node->right, level+1);
}

} // namespace bst

// only here for testing while we implement the gtest module
int main()
{
	std::vector<int> test = {5,7,0,9,3,2,8};
	auto tree = bst::BST<int>(test);
	tree.traverse();
	auto node = tree.search(9);
	if(node)
		std::cout << "node found: " << std::to_string(node->value) << std::endl;
	node = tree.max();
	if(node)
		std::cout << "max node found: " << std::to_string(node->value) << std::endl;
	node = tree.min();
	if(node)
		std::cout << "min node found: " << std::to_string(node->value) << std::endl;
	return 0;
}
